Remove `allowcryptofallback` (#1713)
microsoft/go: eng/doc/fips/README.md

- Running `go version -m` on a binary which uses a system crypto backend now shows the `microsoft_systemcrypto=1` build setting.

- The build-time backend compatibility check now only runs when a crypto package is required for the build.
  - If your app doesn't depend on a crypto package, you may, for example, use `GOOS=linux CGO_ENABLED=0 GOEXPERIMENT=systemcrypto`.
  - If your app doesn't use a crypto package and you make a change that introduces a crypto package dependency, you will only encounter a compatibility check failure after the change. The change may be in your transitive dependencies: for example, depending on a new module that uses `crypto/sha256` may trigger the compatibility check. This is undesirable, but it's necessary to enable flexibility.

- `GOFIPS=0` support has been removed. It now has no effect.

> - `GOEXPERIMENT=boringcrypto` support has been removed.

Reviewer: It never had "support":

> boringcrypto selects the upstream BoringCrypto backend, which is not supported nor compliant

Although unfortunately "support" has multiple meanings... I think adjusting this to fit with the next point makes sense:

```suggestion
- `GOEXPERIMENT=boringcrypto` has been removed.
```

> - `GOEXPERIMENT=allowcryptofallback` has removed. Instead, if it's necessary to opt out from using a system crypto backend, use `GOEXPERIMENT=nosystemcrypto`. This is an internal mechanism that is not intended for use when building a Go application. This document has always recommended against using it, so we anticipate that this change won't affect users of the Microsoft build of Go. Please [contact the maintainers of the Microsoft build of Go](https://github.com/microsoft/go/blob/microsoft/main/SUPPORT.md) if you need to use it so we can understand the scenario and help find a safer alternative.

Reviewer: ```suggestion
- `GOEXPERIMENT=allowcryptofallback` has been removed. Instead, if it's necessary to opt out from using a system crypto backend, use `GOEXPERIMENT=nosystemcrypto`. This is an internal mechanism that is not intended for use when building a Go application. This document has always recommended against using it, so we anticipate that this change won't affect users of the Microsoft build of Go. Please [contact the maintainers of the Microsoft build of Go](https://github.com/microsoft/go/blob/microsoft/main/SUPPORT.md) if you need to use it so we can understand the scenario and help find a safer alternative.
```

- `GOEXPERIMENT=boringcrypto` support has been removed.

> - `GOEXPERIMENT=allowcryptofallback` has been removed. Use `GOEXPERIMENT=nosystemcrypto` to explicitly opt-out from using a system crypto backend.

Reviewer: Worth explicitly preventing someone from possibly interpretating this as a command out of context?

```suggestion
- `GOEXPERIMENT=allowcryptofallback` has been removed. Instead, if it's necessary to opt out from using a system crypto backend, use `GOEXPERIMENT=nosystemcrypto`.
```

> - `GOEXPERIMENT=allowcryptofallback` has been removed. Use `GOEXPERIMENT=nosystemcrypto` to explicitly opt-out from using a system crypto backend.

Reviewer: I'm not sure I agree with removing the bulk of this text. It seems like all that needs to change is the first sentence. Why the rest?

Author: Will re-add the deleted text 👍
